Sources/CombineWamp/PeerWorkflow/Client/WampCaller.swift

```diff
@@ -0,0 +1,40 @@
+import Combine
+import Foundation
+import FoundationExtensions
+
+/// WAMP Caller is a WAMP Client role that allows this Peer to call RPC procedures
+public struct WampCaller {
+    let session: WampSession
+
+    init(session: WampSession) {
+        self.session = session
+    }
+
+    public func call(procedure: URI, positionalArguments: [ElementType]? = nil, namedArguments: [String : ElementType]? = nil)
+    -> Publishers.Promise<Message.Result, ModuleError> {
+        guard let id = session.idGenerator.next() else { return .init(error: .sessionIsNotValid) }
+        let messageBus = session.messageBus
+
+        return session.send(
+            Message.call(.init(request: id, options: [:], procedure: procedure, positionalArguments: positionalArguments, namedArguments: namedArguments))
+        )
+        .flatMap { () -> Publishers.Promise<Message.Result, ModuleError> in
+            messageBus
+                .setFailureType(to: ModuleError.self)
+                .flatMap { message -> AnyPublisher<Message.Result, ModuleError> in
```

Comment:
Do you actually need that inner flatMap here? It seems to me like you could either return or throw. Or do you need flatMap to support the `Empty` case?

Reply:
Problem with throw is that one can't specialise the Error type, to ModuleError.commandError. I can think of a way to simplify this code.